File: java/Group Anagrams.java
// Group Anagrams - 49

import java.util.*;

class Solution {

    //Groups a list of strings into anagrams.
    public List<List<String>> groupAnagrams(String[] strs) {
        
        // Create a map to store the anagrams, where the key is the sorted string and
        // the value is a list of anagrams
        Map<String, List<String>> d = new HashMap<>();

        // Iterate over each string in the input array
        for (String s : strs) {
            // Convert the string to a character array and sort it
            char[] t = s.toCharArray();
            Arrays.sort(t);

            // Convert the sorted character array back to a string to use as a key in the
            // map
            String k = String.valueOf(t);

            // If the key is not already in the map, add it with a new list as its value
            // Then, add the original string to the list of anagrams for this key
            d.computeIfAbsent(k, key -> new ArrayList<>()).add(s);
        }

        // Return a list of the values in the map, which are the lists of anagrams
        return new ArrayList<>(d.values());
    }
}

File: python/Group Anagrams.py
# Group Anagrams - 49

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        
        # Create a hashmap to store the anagrams, with the sorted character counts as keys
        d = defaultdict(list)
        
        # Iterate over each string in the input list
        for s in strs:
            # Initialize a counter array to store the character counts
            cnt = [0] * 26
            
            # Count the occurrences of each character in the string
            for c in s:
                # Use the ASCII value of 'a' as a base to get the index of the character in the counter array
                cnt[ord(c) - ord('a')] += 1
            
            # Use the character counts as a key in the hashmap and append the string to the corresponding list
            d[tuple(cnt)].append(s)
        
        # Return the values of the hashmap as a list of lists, where each sublist contains anagrams
        return list(d.values())